### ISSUE TITLE
UI Simplification: Text element insertion panel - Add Saved Styles

### ISSUE THREAD

**Issue by miina** [swissspidy/web-stories-wp#10113]
## Feature Description
See [Figma](https://www.figma.com/file/dc9xrVlRhQq4KTE9WQzU1R/_Editor---Sprint-Feature-Level-Designs_?node-id=11470%3A451937)

- [x] Add Saved Styles to the Text insertion panel (add this behind a feature flag! To be removed within #10609)
- [x] Clicking on a saved style when no element is selected or non-text element is selected: Insert a new text element with the saved styles
- [x] Clicking on a saved style when text element selected: Apply the saved style to the selected text(s)
- [x] Note: Display the Insertion `+` icon on the Saved Style (when hovering or focused) if clicking on it results in adding a new element
- [x] Display the `+` icon in the panel header for saving a style as well. The icon is disabled if a text element is not selected.
- ~Performance: make sure only the Saved Styles panel re-renders when selecting/de-selecting a text element, and make sure it re-renders only if it actually needs to.~ Added a follow-up issue: https://github.com/GoogleForCreators/web-stories-wp/issues/10705
- [x] Add new tests for the Saved Styles.

## Additional Context

<img width="299" alt="Screenshot 2022-02-15 at 17 48 56" src="https://user-images.githubusercontent.com/3294597/154097706-7ba641e9-42cd-4d44-90ac-219f7f08532b.png">

**Comment by miina** [comment]
To confirm:
- Assuming the Saved Styles still stay in the Style panel as well, is that correct?
- Will clicking on a Saved Style in the Insertion panel insert a text with that style? (Since it's in the insertion panel).

**Comment by agingoldseco** [comment]
We will wait on getting rid of the right panel until the floating bar and other changes are built out, so we'll continue work on organizing the left panel with the new elements for then.

**Comment by andrew-askedall** [comment]
Assuming the Saved Styles still stay in the Style panel as well, is that correct?
-Yes
Will clicking on a Saved Style in the Insertion panel insert a text with that style? (Since it's in the insertion panel).
-Yes

**Comment by agingoldseco** [comment]
Here is the latest version of the panel!

<img width="425" alt="Screen Shot 2022-02-02 at 6 34 28 PM" src="https://user-images.githubusercontent.com/90645922/152255962-84619c6b-f6a2-490d-80a2-da7c28af16cb.png">

We decided to keep the Text Sets section the way it was for now. The only thing we're adding is that Saved Styles section—like in the last round of changes, people can save new ones with the Plus button and click More Styles to see a flyout.

**Comment by miina** [comment]
@agingoldseco Thanks for the update!

On adding the Saved Styles:
Currently, the insertion panel displays consistent content and doesn't really "care" which element is selected or if any is selected.
However, the `+` icon for adding a new Style can work only if a text element is selected while opening the insertion panel, so this means that we could only conditionally display this (only if the user opens the insertion panel while a text element is selected).
Should we still display the `+` icon in the insertion panel as well? Wondering if it might be confusing for the user if it's sometimes there and sometimes not, and it's not related to inserting elements. Thoughts?

**Comment by agingoldseco** [comment]
So I am working on potential ways to allow people to also change text with that panel as opposed to just adding it (can explain further if you're skeptical), but in the case of the + button for saving styles I'd just disable it when text is not selected. Like this:

<img width="384" alt="Screen Shot 2022-02-03 at 10 38 52 AM" src="https://user-images.githubusercontent.com/90645922/152414948-7ba3b7eb-6f3e-4bcc-8f32-1436e50ef6c2.png">

As opposed to it constantly disappearing and reappearing. I used _FG (dark) / Disable_ for this.

**Comment by miina** [comment]
@agingoldseco

Just to confirm that `+` will not change the text, it will _add_ save a new style based on the currently selected Text (as in case of the Style panel). Also, as previously confirmed with Andrew, clicking on a Style will add a new Text with that style directly. See the [answers from Andrew](https://github.com/GoogleForCreators/web-stories-wp/issues/10113#issuecomment-1018773219). So I'm a bit confused now how the user would be able to change the text from insertion panel, could you clarify what you meant by that?

My understanding is currently (based on previous answers from Andrew / new information about `+`):
- This is an insertion panel, meaning that clicking on a style will add a new text element right away with that already applied.
- Clicking on `+` will save a new Style into the panel (if a Text element is selected, otherwise the icon will be disabled).

Let me know if you are thinking something else.

**Comment by swissspidy** [comment]
> but in the case of the + button for saving styles I'd just disable it when text is not selected.

Agreed, this is better from an accessibility perspective. We can make it appear disabled but still focusable by keyboard so that assistive technologies can announce (e.g. read aloud that it's disabled.

> So I'm a bit confused now how the user would be able to change the text from insertion panel, could you clarify what you meant by that?

Perhaps the idea was that clicking on a style will:

* update the currently selected element to have that style if there is one
* create a new text element with that style if no element is selected

Clicking on `+` would continue to save a new style (if a text element is selected, otherwise it's disabled)

@agingoldseco @aaskedall  LMK if I understood this correctly or if you were thinking of something else

**Comment by miina** [comment]
> Perhaps the idea was that clicking on a style will:
> - update the currently selected element to have that style if there is one
> - create a new text element with that style if no element is selected

If that's the case, should the Presets also behave the same way -- apply on the current text if a text is selected? Especially considering that [Presets will be added to the Style panel, too](https://github.com/GoogleForCreators/web-stories-wp/issues/10114).
Would be nice to have a clear indicator for the user then as well -- about if it's going to insert an element or apply style.

**Comment by miina** [comment]
@agingoldseco @andrew-askedall The issue description is now updated based on our sync, tagging you for double checking if I got it all correctly.

**Comment by felipebochehin87** [comment]
Tested against https://stories-qa-wordpress-amp.pantheonsite.io/wp-admin/, using PR #10685.

1. Enabled feature flag Library - Saved Styles
2. Started a new story
3. Added a text and added some style to it
4. Saved style
5. Added saved style - OK
6. Added new text and clicked on the style to set text to the selected style - OK
7. "+" is displayed when hovering over styles
8. Labels working as expected too
